### PR TITLE
Add SSE2 huffman progressive encoding

### PR DESCRIPTION
**Pull request by mayeut**
Derived work from #25

using `time ./jpegtran -outfile /dev/null -progressive -optimise -copy none print_poster_0025.jpg`this gives:

```
Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz

master x86 Apple LLVM version 7.0.2 (clang-700.1.81)
real 0m3.904s
user 0m3.760s
sys  0m0.134s

jcphuff-simd x86 Apple LLVM version 7.0.2 (clang-700.1.81)
real 0m1.926s
user 0m1.784s
sys  0m0.132s

master x86_64 Apple LLVM version 7.0.2 (clang-700.1.81)
real 0m3.454s
user 0m3.295s
sys  0m0.151s

jcphuff-simd x86_64 Apple LLVM version 7.0.2 (clang-700.1.81)
real 0m1.763s
user 0m1.617s
sys  0m0.142s
```

### CODE REVIEW

**Comment by dcommander** [comment]
I am only compensated for my work on libjpeg-turbo if an organization specifically pays for that work or if it benefits VirtualGL or TurboVNC (in which case I can borrow from the "general fund" for those projects.)  I've spent something like 30-40 hours of uncompensated labor thus far working on the baseline Huffman stuff.  It's easy to justify borrowing from the VirtualGL/TurboVNC General Fund to finance that work, since it will directly speed up both of those products.  However, since progressive encoding doesn't have any benefit to VirtualGL or TurboVNC, I'm going to have to secure funding to integrate and test this patch thoroughly.  Hopefully Mozilla or someone else who has a burning need for faster progressive encoding will step forward with money.  Given how much time it's taken me to work on the baseline Huffman encoding patch (despite the fact that your patches were very well written, it always takes time to integrate and test major changes), I can't spare the cycles right now to integrate this.  However, I do very much appreciate the work.  It will land eventually-- just perhaps not in libjpeg-turbo 1.5.

**Comment by dcommander** [comment]
After a cursory look at the patch, I think my main objection (and what caused me to add the "funding needed" tag, which is my default reaction when someone submits a patch that looks like it would require a lot of work to integrate) is that it introduces a lot of new code into jcphuff.c.  The reason why this concerns me is that it creates maintenance difficulties.  The basic design of libjpeg-turbo involves keeping most SIMD-specific code contained in the simd/ directory.  There are only a few exceptions to this:

* In jmemmgr.c, a 16-byte alignment is used when SIMD is enabled, as opposed to the normal 8-byte alignment.
* In jdct.h and jmorecfg.h, the DCT coefficient type is set to short instead of int when SIMD is enabled.
* In jcdctmgr.c, we use scale rather than shift to perform quantization when SIMD is enabled (to avoid overrunning the 16-bit DCT coefficient type.)
* In other files, a check is performed to determine whether SIMD is available for a particular function, and the function pointer is redirected to the SIMD version.

So basically the SIMD functions need to be drop-in replacements for the C functions, and I would like to maintain that architecture for Huffman coding as well.  It appears that there are some fundamental algorithmic differences between the SIMD and non-SIMD versions of the progressive Huffman functions, as well as some algorithmic differences between the 32-bit and 64-bit implementation.  That makes it hard to follow what you are doing, and it raises questions such as:  if those changes to the C portion of the code improve performance, then why can't they be applied to the non-SIMD case as well?  

In general, I would just prefer that the touch points within jcphuff.c be reduced as much as possible.  It's difficult enough to maintain multiple SIMD implementations of the same algorithm.  When each SIMD implementation uses a different algorithm, that becomes untenable.

**Comment by mayeut** [comment]
This uses the same kind of trick as regular huffman SSE2 SIMD encoding in f3a8684cd1c28e557d394470962a7a224c76ddbc .
That is:
1. Prepare data & compute 64 bits usable by `bsf`. (Here in assembly). 
2. Replace the index loop by a `bsf`loop to skip whole bunch of zero at a time. (Here in C).

About the 32-bit/64-bit difference: well it's simply that `bsf` can work with 64-bit register in 64-bit implementation. In the 32-bit implementation, the loop is done twice (unrolled with some glue between the loops). The same is done in the regular huffman SSE2 SIMD encoding.

Regarding implementing the whole functions in assembly as drop-in replacements, it's certainly doable, however, I think it will bring much more complexity. The functions `emit_...` have dual behavior here depending on `gather_statistics` value. This means twice the code in assembly for full performances. Here I let the compiler handle that (maybe badly but more easily fixable using inlined functions) which I think ease maintenance.

Improving the non-SIMD case: well, there's probably something to do here as well.
However, for `encode_mcu_AC_first`, it seems hard to get something done that will be similar to the SIMD optimization. We're really in a situation like the one from regular huffman encoding with the dual behavior added.
For `encode_mcu_AC_refine`, there's already a data preparation pass in the non-SIMD code and thus, something can probably be done so that this be rewritten to provide a drop-in replacement for data preparation.

There's a compromise to be made between performances & maintainability. I chose one that I deemed maintainable (full assembly would probably have been more performant but unmaintainable). We can drop a bit of performance & get something even more maintainable.

**Comment by dcommander** [comment]
Perhaps I was misreading the code, then.  Do the underlying routines (jsimd_encode_mcu_AC_first_prepare_sse2(), jsimd_encode_mcu_AC_refine_prepare_sse2()) accept identical inputs and produce identical outputs on both 32-bit and 64-bit?

The ultimate goal is to provide a common interface that can be easily leveraged by other (non-x86) architectures as well-- for instance, if someone wanted to develop a NEON or MIPS or AltiVec version of these same functions, they might choose a different approach, because those architectures have different constraints.  I don't really object to introducing hybrid C/assembly functions.  I just think it may make more sense to move as much of that code as possible into the simd/ directory.  I'm happy to do that work.  I just need to get a better understanding of what's going on-- and perhaps a sense of whether the existing approach is easily portable to, for instance, NEON.

**Comment by mayeut** [comment]
`jsimd_encode_mcu_AC_first_prepare_sse2` and `jsimd_encode_mcu_AC_refine_prepare_sse2` accept the same input on both 32-bit and 64-bit. The output is a little bit different. In 32-bit, it outputs 2 (resp. 4) 32-bit values. In 64-bit,  it outputs 1 (resp. 2) 64-bits values. Memory content of those values are identical. Those values are called `zerobits` in the C code (bit loop indices) and `signbits` as another output specific to `jsimd_encode_mcu_AC_refine_prepare_sse2` (precomputed value for `((*block)[jpeg_natural_order_ss[k]] < 0) ? 0 : 1;`).

I don't know about MIPS and AltiVec so I won't make assumptions.
For NEON, the same approach is easily portable. The only thing is that the output for `zerobits` won't be the same. x86/x64 has `bsf` instruction but ARM doesn't have this. ARM has `clz`. It's almost the same but the loop needs to be reverted (ARM also has `rbit` instruction which makes this easy). This is what's done in #47 for regular huffman encoding.
In other words, the C code will be clobbered with more `#if defined(HAVE_JCPHUFF_BUILTIN_CLZL)` or a macro/inlined function can be written once for the right operation. For the shift, this can be defined as a macro 
```
#if defined(HAVE_JCPHUFF_BUILTIN_CLZL)
#define BITLOOPSHIFT(x, shift) ( (x) <<= (shift) )
#else
/*
 * GCC: HAVE_JCPHUFF_BUILTIN_CTZL
 * MSVC x64: HAVE_JCPHUFF_BITSCANFORWARD64
 * MSVC x86: HAVE_JCPHUFF_BITSCANFORWARD
 */
#define BITLOOPSHIFT(x, shift) ( (x) >>= (shift) )
#endif
```

**Comment by dcommander** [comment]
I took another look at this, and there are just too many problems with the layout of the code for it to be acceptable in its current state.  I am open to suggestions for how to fix it.  Here are the problems as I see them:

* Any significant SIMD-specific code needs to reside under simd/.  We can't make the assumption that your approach will be the best approach on all SIMD architectures-- in fact, you couldn't even use the same approach with i386 and x86-64, so that doesn't give me confidence that your approach is generically applicable to other instruction sets.  If it proved necessary to use a different approach with, for instance, AltiVec, then we'd be faced with a bunch of very ugly #ifdefs in jcphuff.c.  I attempted to refactor your code so that the SIMD-specific code all resides under simd/, but this was problematic as well, because there is so much reuse of code from jcphuff.c that I would have had to clone most of that file into the simd/ directory or break out the common sections of it into a separate file.  Neither one is a good proposition if I want to ensure maintainability of the code.
* I would strongly prefer that intrinsics not be used on x86.  There are just too many compiler differences between the various x86 platforms, as evidenced by the #ifdefs in your code.  A possible compromise would be to abstract these operations using macros, as you described above.
* If a SIMD function for one particular platform has different inputs or outputs than the same function for another platform, then those differences need to be abstracted in the stub file (jsimd_x86_64.c and jsimd_i386.c, in this case), and a common interface needs to be presented to the C portion of the library.  This is again so that implementations for other SIMD architectures will not be constrained by your particular approach.

I am not saying that you have to rewrite the whole thing using assembly.  I'm just saying that we need to open a discussion as to how best to refactor it so that the above concerns are addressed.

One thing to note is that jcphuff.c was never optimized like jchuff.c was, so it still has the old IJG implementation of emit_bits(), etc. rather than my optimized version (EMIT_BITS().)  This is why I mentioned optimizing the C algorithm first.  You're only using SIMD for the preparation step.  Would it not be straightforward to implement a C version of the *prepare() functions?  That way, the same code structure could be reused for both SIMD and non-SIMD cases.  This would eliminate the maintainability concerns.  Also, it's quite likely that emit_bits() and emit_symbol() could be straightforwardly implemented in assembly, perhaps even borrowing the some of the implementation from baseline Huffman encoding.  This would allow the entire inner loop for the SIMD implementation to be implemented using assembly, and that would make it very straightforward to wrap the SIMD implementation, as we did with jchuff.c.

To summarize the possible approaches, from easiest to hardest (please let me know if I've gotten something wrong here):
1. Implement a C version of the prepare*() functions, then use the same algorithm you're using with SIMD for the non-SIMD case as well.  I would also want the prepare*() functions to either present a common interface to the C code, or I would need a good explanation as to why you feel the 32-bit interface would be generically applicable to all 32-bit platforms, and likewise for the 64-bit interface.  There is precedent for separate 32/64-bit code in jchuff.c, so if this is a similar case, then I'm OK with that, as long as the code split is for all 32-bit and 64-bit platforms and is not specific to x86.
2. Optimize the C code first, which would probably require a lot of my input, since I wrote the optimizations for jchuff.c.  Then it should be more straightforward to port the inner loops to assembly.
3. Maintain the existing unoptimized C implementation but port the entirety of the inner loops for the SIMD implementation to assembly (mainly, you'd need emit_symbol(), emit_bits(), and emit_eobrun().  The rest should be straightforward.)

Honestly, it seems like (1) may be the most straightforward approach, assuming that it doesn't regress performance for the non-SIMD case.

Although 2X speedup is very compelling, it unfortunately still isn't compelling enough for me to invest much more time into this unless someone is paying for that time.  For my own purposes, even with a 2X speedup, progressive is still 3-6x slower than baseline, and it only compresses about 5-15% better.  My other projects deal with remote real-time display of 3D applications, so better compression is always wanted for low-bandwidth networks, but 5-15% better compression for a 3-6x hit in performance (relative to baseline) isn't a good trade-off.  I've approached the mozjpeg developers to see if they are interested in funding this, but I haven't heard back.  This enhancement speeds up mozjpeg by about 25-30%.

**Comment by mayeut** [comment]
I started to work on something that looks like (1) which, as you said is the easier approach.
I stopped working on that to check the ARM64 PR.

For now, I started with `encode_mcu_AC_refine` which already has a prepare pass. It seems promising, even in the non-SIMD case (+17% for clang x64, +45% for gcc 5.2.0 which was quite slower than clang this time around). SSE2 on clang gives +69%.

As I said in the previous comment, the reason for a difference between 32bits/64bits is because of register width  (maybe that was not clear), like what's done in `jchuff.c` to hold onto more data & reduce checks for flushing if I understood correctly what you did.

I also started to abstract intrinsics in an inlined function (when missing, just use plain C code, not as bad as I thought, still better than original implementation)

The only reason I see for this unified `encode_mcu_AC_refine` not to be possible would be to take into account different DCTSIZE/DCTSIZE2. I guess there's a reason the checks exist in `jsimd_can_*`. If DCTSIZE2 has no boundaries, then it might be hard to get something unified here, at least in this form. Less than 64 is not a problem, more than 64 can become tedious.

The work I've done so far with this function can still be improved with your 2nd option of optimizing the C code (mainly `emit_*` as you said)

**Comment by dcommander** [comment]
Cool. That sounds very promising. Most of the SIMD algorithms will only work with an 8x8 DCT, so it's safe to make that assumption here as well.

**Comment by mayeut** [comment]
Sorry I think was not clear on the 8x8 DCT. I meant that the `prepare` C version which allows to unify SIMD/non-SIMD versions also relies on DCT not being larger than 8x8 (1 64-bit / 2 32-bit register used to hold if something needs to be done for a each coefficient & 1 64-bit / 2 32-bit register used to hold the sign of each coefficient). If DCT size is more than 8x8, the unified C implementation can't work (as of now). Anything less than 8x8 works though.

**Comment by dcommander** [comment]
DCT size should never be anything other than 8x8 for compression. SmartScale was a proposal by the current maintainer of IJG, which implements other DCT sizes, but his proposal was rejected by the standards committees, and my own testing did not reveal any particular use for it (gorey details here:  http://www.libjpeg-turbo.org/About/SmartScale)

**Comment by mayeut** [comment]
I commited a new branch mayeut@20f8aa9 which does only optimization for `encode_cu_AC_refine`.
As stated before, it's the only function which already has a prepare pass. This prepare pass now has its own function in C & SSE2.
All performances for x86 & x64 are in the commit log.

I also tested it on the ARM64 Cavium Thunder X (no SIMD for ARM64 yet, only the new C path), this gives +40% performance increase.

@dcommander please tell me if you wan't me to open a new PR with only this or just forget about all that was previously done in this PR & switch it to the new commit.

**Comment by dcommander** [comment]
I'll let you know.  I'm going to be busy preparing libjpeg-turbo 1.5 over the next few weeks, but I definitely want to get this feature into libjpeg-turbo 1.6.

**Comment by mayeut** [comment]
Given the timings I got for the C version with clang aarch64 on iPhone 5S, I checked what was the result for armv7 on the same hardware.
This gives +6% so no regressions here.

**Comment by dcommander** [comment]
Thanks for the effort.  I'll give this a more thorough look after the 1.5 release.  Please keep everything under this pull request for now.

**Comment by dcommander** [comment]
One thing we're blocking on for this is the fact that mozjpeg will also benefit from it, so I will be working with them to integrate these enhancements in a way that can benefit both projects.

**Comment by felixbuenemann** [comment]
@mayeut Mozjpeg already has a [similar optimization](https://github.com/mozilla/mozjpeg/blob/ec333d5/jcphuff.c#L512) for `encode_mcu_AC_first`, but it also seems to handle non-2s complement systems and gets rid of the `if (temp == 0)` check. It seems to be slightly faster than your version, too, so you might want to take a look at it.

I measured the performance when [applied to mozjpeg](https://github.com/felixbuenemann/mozjpeg/tree/mayeut-jcphuff-acrefine) and it is less extreme due to the before-mentioned optimization already being present and because by default progressive scan optimization and trellis quantization are enabled, which take a lot of cpu time. Encoding a 10000x10000 jpeg with `-revert -progressive` in mozjpeg took ~30% less time, with the default settings encoding took ~10% less time.

**Comment by dcommander** [comment]
I was able to get much more than a 30% speedup by applying this patch to mozjpeg 3.1, but I am working with them to study this in more detail. Stand by.

**Comment by dcommander** [comment]
Confirmed that, at least for x86-64, this commit still speeds up compression with mozjpeg 3.1 by about 15% across the board.  Pending funding, I would like to upstream the optimizations that mozjpeg has already done, add these optimizations to them, and do any necessary cleanup, integration, and testing work.

**Comment by dcommander** [comment]
@mayeut I am ready to proceed with integrating this, but I'm not sure where we left off in terms of how best to integrate it.  Can you please rebase your latest work (the code that has both C and SIMD implementations of the _prepare() functions) onto dev so I can more easily review it?  Thanks.

**Comment by mayeut** [comment]
@dcommander, this is now rebased onto dev.

Updated the commit message to show improvement against current dev branch:

This commit adds C & SSE2 optimizations for encode_mcu_AC_refine
function of Huffman progressive encoding.

The image used for testing can be retrieved from the page
https://blog.cloudflare.com/doubling-the-speed-of-jpegtran/

All timings done on `Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz`
clang version is `Apple LLVM version 9.0.0 (clang-900.0.39.2)`
gcc-5 version is `gcc-5 (Homebrew GCC 5.5.0) 5.5.0`
gcc-7 version is `gcc-7 (Homebrew GCC 7.2.0) 7.2.0`

Here are the results in comparison to
libjpeg-turbo@3c54642 using `time ./jpegtran -outfile /dev/null -progressive -optimise -copy none print_poster_0025.jpg`

C
clang x86_64 +7%
gcc-5 x86_64 +30%
gcc-7 x86_64 +33%
clang i386 +0%
gcc-5 i386 +24%
gcc-7 i386 +23%

SSE2
clang x86_64 +42%
gcc-5 x86_64 +53%
gcc-7 x86_64 +64%
clang i386 +35%
gcc-5 i386 +46%
gcc-7 i386 +49%

**Comment by dcommander** [comment]
Thanks much.  I'll look at it early next week.

**Comment by mayeut** [comment]
The improvement in throughput was against the preview image of `print_poster_0025.jpg`
When applied to the full-size image, the improvement becomes:

C
clang x86_64 +15%
gcc-5 x86_64 +44%
gcc-7 x86_64 +53%
clang i386 +6%
gcc-5 i386 +36%
gcc-7 i386 +32%

SSE2
clang x86_64 +53%
gcc-5 x86_64 +70%
gcc-7 x86_64 +93%
clang i386 +52%
gcc-5 i386 +64%
gcc-7 i386 +63%

**Comment by mayeut** [comment]
@dcommander, have you been able to look at this ? do you need me to make some modifications ?

**Comment by dcommander** [comment]
Haven't been able to look at it yet.  I fell down another rabbit hole.  Hope to get to this later today or tomorrow.

**Comment by dcommander** [comment]
The latest code is unfortunately slower than the previous, by about 15% on average.  Any ideas why?

**Comment by mayeut** [comment]
I just rebased on latest dev & started the same bench as a couple weeks ago. I've got the same results here.
Since there have been different approach to optimize this & different rebase, can you be a bit more explicit on "the previous" so that I can check exactly what could be going on ?

**Comment by dcommander** [comment]
I have to leave for the day, but I'll try to re-find the exact code I was testing tomorrow or Monday.

**Comment by dcommander** [comment]
@mayeut I pushed a [new branch](https://github.com/libjpeg-turbo/libjpeg-turbo/tree/dev.jcphuff_simd_working) with the code base that performs as I expect.  I can't readily tell why the performance of that branch is different from the code currently in this PR.

**Comment by mayeut** [comment]
@dcommander,
Based on the branch you pushed, I can tell that the code in this PR is not based on the same commit:
- your branch is based on mayeut@70438c999036d766520971fa618f0e7efe34e489
- the PR has been rebased using mayeut@4b19aa7471bcdae51e6c4f71d123fd579ce6c2bb

Performance differences (a couple years back) between those 2 commits were roughly what you're seeing.
The reasons why I rebased using the last one is because of these comments: https://github.com/libjpeg-turbo/libjpeg-turbo/pull/46#issuecomment-172136745 and https://github.com/libjpeg-turbo/libjpeg-turbo/pull/46#issuecomment-172136745

Mostly, and it was a very good point in the first comment, mayeut@70438c999036d766520971fa618f0e7efe34e489 had code duplication in `jcphuff.c` with few differences depending on SIMD usage or not. This comment lead to mayeut@4b19aa7471bcdae51e6c4f71d123fd579ce6c2bb implementation but, in that one, only `encode_mcu_AC_refine` has been optimized. I might be able to check if I can add `encode_mcu_AC_first` over the week-end.

**Comment by dcommander** [comment]
I apologize if my comments led to confusion.  My main concern was that, per the design principles of libjpeg-turbo, the behavior of the C code needs to be the same regardless of whether SIMD extensions are being used.  Thus, my only real objection to the initial code-- what's now in the dev.jcphuff_simd_working branch-- is that it used completely different C routines as a front end for the SIMD extensions.

It seems like I may be able to merge the two approaches.  I'll work on that this week.  I don't want you to have to do any additional work on this, and I'm really sorry it's been such a hassle thus far.

**Comment by mayeut** [comment]
Well, I was too curious about performances when implementing a `prepare` step for `encode_mcu_AC_first` so I went for it anyway (it's in a different commit though, I did not squash both commits). It's better than what I expected (at least on my laptop) when running only the C code. It's worth mentioning that it might not be the case on other platforms or even other x86 CPUs. The `encode_mcu_AC_first` did not have a `prepare` step so performance regression risks are higher for this function than for `encode_mcu_AC_refine`.

Now, both `encode_mcu_AC_first` and `encode_mcu_AC_refine` are using the same C routines as a front end for the SIMD extensions.

Some more detailed figures than in the commit log:
```
version tested: time, speedup against dev
clang 64 dev: 3.141174
clang 64 c: 2.627296, 1.195592
clang 64 sse2: 1.745368, 1.799720
gcc5 64 dev: 4.342426
gcc5 64 c: 2.409172, 1.802456
gcc5 64 sse2: 1.680494, 2.584018
gcc7 64 dev: 3.773040
gcc7 64 c: 2.389831, 1.578790
gcc7 64 sse2: 1.692865, 2.228790
clang 32 dev: 3.279643
clang 32 c: 3.122143, 1.050446
clang 32 sse2: 1.913962, 1.713536
gcc5 32 dev: 4.439099
gcc5 32 c: 2.777694, 1.598124
gcc5 32 sse2: 1.895325, 2.342132
gcc7 32 dev: 4.343601
gcc7 32 c: 2.863487, 1.516892
gcc7 32 sse2: 1.843910, 2.355647
```

**Comment by dcommander** [comment]
Cool!  I'm working on integrating this offline, since I need to reformat the code anyhow.  I'll pull in the new commit and squash it.  Thanks again for your hard work!

**Comment by dcommander** [comment]
Integrated.  Note that master is now the working branch for 2.0 beta1.  The only remaining thing I need to do before releasing the beta is figure out why Travis won't build PRs.

**Comment by dcommander** [comment]
@mayeut Unfortunately I discovered a UBSan issue in this code when 12-bit data precision is used (`-DWITH_12BIT=1`).  Can you investigate?

**Comment by mayeut** [comment]
@dcommander, do you have a reproducer for this ?
